### PR TITLE
FINERACT-1678-loan-cob-drop-softlocking-use-lastclosedbusinessdate

### PR DESCRIPTION
**Pull request by ruchiD**
## Description
Dropped applying softlock for loan cob.
Modified Loancobapi filter and worker apply loan tasklet.
Modified tests.

Describe the changes made and why they were made.

Ignore if these details are present on the associated [Apache Fineract JIRA ticket](https://github.com/apache/fineract/pull/1284).

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [ ] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [ ] Create/update unit or integration tests for verifying the changes made.

- [ ] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [ ] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/legacy-docs/apiLive.htm with details of any API changes

- [ ] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

### CODE REVIEW

File: fineract-provider/src/main/java/org/apache/fineract/cob/loan/ApplyLoanLockTasklet.java [hunk 1]
```diff
@@ -69,18 +66,11 @@ public RepeatStatus execute(@NotNull StepContribution contribution, @NotNull Chu
...
-        Map<Long, LoanAccountLock> alreadySoftLockedAccountsMap = accountLocks.stream()
-                .filter(e -> LockOwner.LOAN_COB_PARTITIONING.equals(e.getLockOwner()))
-                .collect(Collectors.toMap(LoanAccountLock::getLoanId, Function.identity()));
-
         List<Long> alreadyLockedByChunkProcessingAccountIds = accountLocks.stream()
                 .filter(e -> LockOwner.LOAN_COB_CHUNK_PROCESSING.equals(e.getLockOwner())).map(LoanAccountLock::getLoanId).toList();
 
-        List<Long> toBeProcessedLoanIds = new ArrayList<>(alreadySoftLockedAccountsMap.keySet());
-
-        loanLockingService.upgradeLock(toBeProcessedLoanIds, LockOwner.LOAN_COB_CHUNK_PROCESSING);
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/cob/loan/ApplyLoanLockTasklet.java:81]: we should not remove applying hard lock on the loans!

File: fineract-provider/src/test/java/org/apache/fineract/cob/loan/FetchAndLockLoanStepDefinitions.java [hunk 2]
```diff
@@ -1,178 +0,0 @@
-/**
```
Comment (adamsaghy) [inline fineract-provider/src/test/java/org/apache/fineract/cob/loan/FetchAndLockLoanStepDefinitions.java:1]: Why did you remove the whole test?

File: fineract-provider/src/test/java/org/apache/fineract/infrastructure/jobs/filter/LoanCOBApiFilterTest.java [hunk 3]
```diff
@@ -265,6 +263,7 @@ void shouldProceedWhenRescheduleLoanIsNotLockedAndNotBehind() throws ServletExce
         verify(filterChain, times(1)).doFilter(request, response);
     }
 
+    @Disabled("Not relevant after removing loan soft lock for cob processing")
```
Comment (adamsaghy) [inline fineract-provider/src/test/java/org/apache/fineract/infrastructure/jobs/filter/LoanCOBApiFilterTest.java]: if it is not relevant, please remove it

File: fineract-provider/src/test/java/org/apache/fineract/infrastructure/jobs/filter/LoanCOBApiFilterTest.java [hunk 4]
```diff
@@ -299,16 +297,15 @@ void shouldRunInlineCOBAndProceedWhenLoanIsBehind() throws ServletException, IOE
...
         given(result.getId()).willReturn(2L);
+        given(result.getLastClosedBusinessDate()).willReturn(businessDate.minusDays(2));
         given(request.getPathInfo()).willReturn("/v1/loans/2?command=approve");
         given(request.getMethod()).willReturn(HTTPMethods.POST.value());
         given(loanAccountLockService.isLoanHardLocked(2L)).willReturn(false);
-        given(loanAccountLockService.isLoanSoftLocked(2L)).willReturn(true);
         given(fineractProperties.getQuery()).willReturn(fineractQueryProperties);
         given(fineractQueryProperties.getInClauseParameterSizeLimit()).willReturn(65000);
-        given(loanRepository.findAllNonClosedLoansBehindOrNullByLoanIds(
+        given(loanRepository.findAllNonClosedLoansBehindByLoanIds(
```
Comment (adamsaghy) [inline fineract-provider/src/test/java/org/apache/fineract/infrastructure/jobs/filter/LoanCOBApiFilterTest.java:287]: why did you change this?

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/jobs/filter/LoanCOBApiFilter.java [hunk 5]
```diff
@@ -184,17 +184,12 @@ private List<Long> getGlimChildLoanIds(Long loanIdFromRequest) {
...
-    }
-
     private boolean isLoanHardLocked(List<Long> loanIds) {
         return isLoanLocked(loanIds, true);
     }
 
     private boolean isLoanLocked(List<Long> loanIds, boolean isHardLock) {
-        return isHardLock ? loanIds.stream().anyMatch(loanAccountLockService::isLoanHardLocked)
-                : loanIds.stream().anyMatch(loanAccountLockService::isLoanSoftLocked);
+        return loanIds.stream().anyMatch(loanAccountLockService::isLoanHardLocked);
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/infrastructure/jobs/filter/LoanCOBApiFilter.java:188]: you might wanna remove the `isHardLock` parameter. It is not needed anymore.

File: fineract-provider/src/test/java/org/apache/fineract/cob/loan/FetchAndLockLoanStepDefinitions.java (hunk 2 above)
Comment (ruchiD) [reply fineract-provider/src/test/java/org/apache/fineract/cob/loan/FetchAndLockLoanStepDefinitions.java:1]: This was testing LoanLockTasklet execute method. LoanLockTasklet was no more required as it was applying softlock for loans.

File: fineract-provider/src/test/java/org/apache/fineract/infrastructure/jobs/filter/LoanCOBApiFilterTest.java (hunk 4 above)
Comment (ruchiD) [reply fineract-provider/src/test/java/org/apache/fineract/infrastructure/jobs/filter/LoanCOBApiFilterTest.java:287]: LoanCOBApiFilter isLoanBehind Method calls findAllNonClosedLoansBehindByLoanIds and not findAllNonClosedLoansBehindOrNullByLoanIds. 
The test was passing earlier as isLoanSoftLock and isLoanBehind were in OR condition ((isLoanSoftLocked(loanIds) || isLoanBehind(loanIds)) and test was setting softlock so isLoanBehind condition was not getting checked.
Once isSoftLock check was removed test started failing...

Comment (adamsaghy) [reply fineract-provider/src/test/java/org/apache/fineract/infrastructure/jobs/filter/LoanCOBApiFilterTest.java:287]: The inline COB job in the other hand is using the `findAllNonClosedLoansBehindOrNullByLoanIds` it should have been called...

I guess we just need to check further this test case was about to check what particular test case.
